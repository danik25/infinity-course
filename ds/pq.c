/****************************************************************
*                <<<<<<<< ASSIGNMENT NAME >>>>>>>>              *
* Related files: Priority.c, Priority_test.c                    *
* #Version: V 1.2                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Or Stopiky                                          *
* Last update: 17/04/19                                         *
****************************************************************/ 
#include <stdlib.h> /*free*/
#include <assert.h> /*assert*/

#include "sorted_list.h"
#include "pq.h"

struct pqueue {
    
    sorted_list_t *sorted_list;
    
};

pqueue_t *PQCreate(int (is_before)(void *data1, void *data2, void *param),
                                                                  void *param)
{
    pqueue_t *s_q = (pqueue_t *) malloc(sizeof(pqueue_t));
    sorted_list_t *s_list = NULL;

    if(!s_q)
    {
        return (NULL);
    }
    
    s_list = SortCreate(is_before, param);

    if(!s_list)
    {
        free(s_q);
        return (NULL);
    }

    s_q->sorted_list = s_list;
    return (s_q);
}


void PQDestroy(pqueue_t *queue)
{
    assert(queue);

    SortDestroy(queue->sorted_list);
    free(queue);
    queue = NULL;
}

int PQEnqueue(pqueue_t *queue, void *data)
{
    assert(queue);

    if(SortInsert(queue->sorted_list, data))
    {
        return(0);
    }
    return (1);
}

void *PQDequeue(pqueue_t *queue)
{
    assert(queue);

    return (SortPopFront(queue->sorted_list));
}

void *PQPeek(const pqueue_t *queue)
{
    assert(queue);

    if(PQIsEmpty(queue))
    {
        return (NULL);
    }

    return (SortGetData(SortBegin(queue->sorted_list)));
}

int PQIsEmpty(const pqueue_t *queue)
{
    assert(queue);

    return (SortIsEmpty(queue->sorted_list));
}

size_t PQSize(const pqueue_t *queue)
{
    assert(queue);

    return (SortSize(queue->sorted_list));
}

void PQClear(pqueue_t *queue)
{
    assert(queue);

    while(!PQIsEmpty(queue))
    {
        PQDequeue(queue);
    }
}

void *PQRemove(pqueue_t *queue, int(* is_match_func)(void *data, void *arge),
                                                                    void *data)
{
    sorted_iter_t node = NULL;
    void * removed_data = NULL;
    node = SortFind(queue->sorted_list, SortBegin(queue->sorted_list), 
        SortEnd(queue->sorted_list), is_match_func, data);
    if(node != SortEnd(queue->sorted_list))
    {
        removed_data = SortGetData(node);
        SortErase(node);
    }
    return (removed_data);
}