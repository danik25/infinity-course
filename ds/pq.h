#ifndef _PRIORITY_QUEUE_
#define _PRIORITY_QUEUE_

#include <stddef.h> /* size_t */



/****************************************************************
*                <<<<<<<< ASSIGNMENT NAME >>>>>>>>              *
* Related files: Priority.c, Priority_test.c                    *
* #Version: V 1.2                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Or Stopiky                                          *
* Last update: 17/04/19                                         *
****************************************************************/  
/* API DESCRIPTION:
 *
 * Priority queues are a type of container adaptors, specifically designed such 
 * that its first element is always the greatest of the elements it contains, 
 * according to some strict ordering criterion supplied by the user.
 *
 *
 * 
 */
 

typedef struct pqueue pqueue_t;


/* Description: 
            Creates an empty priority queue.
 *                  
 * Params: 
 *      Is_before - a user supplied function that will determine the 
 *      priority for sorting the elements of the queue.
 *      Param - an additional paramater for the use of the user.      
 *        
 * Return Value: 
 *        Pointer to the newly created queue.        
*/
pqueue_t *PQCreate(int (is_before)(void *data1, void *data2, void *param),
                                                                  void *param);


/* Description:
 *          Free all allocated memory of the queue and destroy the queue. 
 *
 * Params:
 *      queue - the priority queue to be destroyed.  
 *        
 *        
 * Return Value: void       
*/
void PQDestroy(pqueue_t *queue);



/* Description: 
 *          Unserts an element to the queue in a location determined by its
 *          priority.
 *
 * Params: 
 *       queue - the queue to insert to.    
 *       data - the data to insert. 
 *        
 * Return Value: 
 *         status - success(0) or failure(1).
*/
int PQEnqueue(pqueue_t *queue, void *data);



/* Description:
 *          Removes the element with the highest priority from the queue.
 *          Time complexity: O(n)
 *
 * Params:
 *     queue - the queue to remove from.
 *               
 * Return Value: 
 * 
 *      The data of the removed element.
 *      If the queue is empty - returns NULL.     
*/
void *PQDequeue(pqueue_t *queue);



/* Description:
 *          Shows the data of the element with the highest priority.
 *
 * Params:
 *      queue - the queue to look at.  
 *             
 * Return Value: 
 *          The data of the highest priority element.
 *          If the queue is empty - returns NULL.
*/
void *PQPeek(const pqueue_t *queue);



/* Description:
 *          Checks whether the queue is empty or not.
 *
 * Params:
 *      queue - the queue to look in.
 *            
 * Return Value: 
 *      Boolean (TRUE(1) / FALSE(0)).           
*/
int PQIsEmpty(const pqueue_t *queue);



/* Description:
 *          Counts number of element in the queue.
 *
 * Params:
 *     queue - the queue to count.   
 *        
 * Return Value: 
 *      Number of elements.
 *          
*/
size_t PQSize(const pqueue_t *queue);



/* Description:
 *          Free all allocated memory of the queue's elements leaving an
 *          empty queue without destroying it.
 *
 * Params:
 *      queue - te queue to clear.
 *        
 *        
 * Return Value: 
 *              void.
*/
void PQClear(pqueue_t *queue);



/* Description:
 *          Removes a specific element from the queue that matches a user 
 *          specified data.
 * 
 * Params:
 *     is_match_func - used to compare the user data with the queue elements and
 *                     find if a match exist (boolean function).    
 *                     data - the user supplied data for comarison.
 *        
 * Return Value: 
 *              The data of the removed element or NULL if element was not 
 *              found.
*/
void *PQRemove(pqueue_t *queue, int (is_match_func)(void *data, void *args),
                                                                    void *args);
                                                                    
#endif /* _PRIORITY_QUEUE_ */