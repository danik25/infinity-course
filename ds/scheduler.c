/****************************************************************
*                   Scheduler                                   *
* Related files: scheduler.c, scheduler_test.c scheduler.h      *
*                task.c, task.h uid.h, uid.c                    *
* #Version: V 1.3                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Moshe Greenberg                                     *
* Last update: 21/04/19                                         *
****************************************************************/
#include <time.h> /* time_t*/
#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/

#include "scheduler.h"
#include "pq.h" /* pqueue_t */

#define UNUSED(x) (void)(x)

/*compare function*/
int IsBefore(void *data1, void *data2, void *param);
/*compare function*/
int IsMatch(void *data, void* arg);

UID_t g_bad_uid;

struct sched {

    pqueue_t *scheduler;
	int kill_flag; /* flag to stop the run() function. 1 = stop, 0 = run */
};

sched_t * SCHCreate()
{
    sched_t *sched = (sched_t *) malloc(sizeof(sched_t));
    if(!sched)
    {
        return(NULL);
    }
    sched->scheduler = PQCreate(IsBefore, NULL);
    if(!sched->scheduler)
    {
        free(sched);
        return(NULL);
    }
    sched->kill_flag = 0;

    return (sched);
}

void SCHDestroy(sched_t *sched)
{
    assert(sched);
    while(!PQIsEmpty(sched->scheduler))
    {
        TaskDestroy(PQDequeue(sched->scheduler));
    } 
    PQDestroy(sched->scheduler);
    free(sched);
    sched = NULL;
}

UID_t SCHAdd(sched_t *sched, task_function_t act_func, void *param,
                                                          size_t interval)
{
    task_t *task = NULL;

    assert(sched);
    assert(act_func);

    task = TaskCreate(act_func, param, interval);
    if(!task)
    {
        return(g_bad_uid);
    }
    PQEnqueue(sched->scheduler, task);

    return(TaskGetUID(task));
}

int SCHRemove(sched_t *sched, UID_t task_id)
{
    void *removed = NULL;

    assert(sched);

    removed = PQRemove(sched->scheduler, IsMatch, &task_id);

    if(!removed)
    {
        return(1);
    }
    free(removed);
    removed = NULL;
    return(0);
}

int SCHRun(sched_t *sched)
{
    size_t i = 0;
    int status = 0;
    void *tmp_deq = NULL;
    sched->kill_flag = 0;
    for(i = 0; i < SCHSize(sched); ++i)
    {
        TaskTimeUpdate(PQPeek(sched->scheduler));
        tmp_deq = PQDequeue(sched->scheduler);
        PQEnqueue(sched->scheduler, tmp_deq); 
    }
    while(!sched->kill_flag)
    {
        if(SCHIsEmpty(sched))
        {
            return(0);
        }
        sleep(TaskGetTimeToRun(PQPeek(sched->scheduler)) - time(NULL));
        status = TaskRun(PQPeek(sched->scheduler));
        if(status == 1 || status == -1)
        {
            SCHRemove(sched, TaskGetUID(PQPeek(sched->scheduler)));
        }
        else
        {
            TaskTimeUpdate(PQPeek(sched->scheduler));
            tmp_deq = PQDequeue(sched->scheduler);
            PQEnqueue(sched->scheduler, tmp_deq);  
        }
    }
    return (1);
}

void SCHStop(sched_t *sched)
{
    assert(sched);
    sched->kill_flag = 1;
}

size_t SCHSize(const sched_t *sched)
{
    assert(sched);
    return (PQSize(sched->scheduler));
}

int SCHIsEmpty(const sched_t *sched)
{
    assert(sched);
    return(PQIsEmpty(sched->scheduler));
}

/****************************************************************/

/*compare function*/
int IsBefore(void *data1, void *data2, void *param)
{
    UNUSED(param);
    if(TaskGetTimeToRun(data1) < TaskGetTimeToRun(data2))
    {
        return (1);
    }
    return (0);
}

/*compare function*/
int IsMatch(void *data, void* arg)
{
    return (UIDIsSame(TaskGetUID(data), *(UID_t*)arg));
}