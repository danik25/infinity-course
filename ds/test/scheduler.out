/home/xyz/git/dani-benjamin/ds/scheduler.out